### PR TITLE
Update Horde_Text_Diff_Engine_Native, fix array pointer issues

### PR DESCRIPTION
**Pull request by dvz**
fix #2797
https://github.com/mybb/mybb/issues/2743#issuecomment-341786421

based on patches shared on https://core.trac.wordpress.org/ticket/41526

### CODE REVIEW

File: inc/3rdparty/diff/Diff/Engine/Native.php [hunk 1]
```diff
@@ -196,18 +196,30 @@ protected function _diag ($xoff, $xlim, $yoff, $ylim, $nchunks)
                     continue;
                 }
                 $matches = $ymatches[$line];
-                foreach ($matches as $y) {
+                reset($matches);
+                while ($y = current($matches)) {
```
Comment (euantorano) [inline inc/3rdparty/diff/Diff/Engine/Native.php]: Reading [the docs for `current`](http://php.net/manual/en/function.current.php) and the return value, shouldn't we explicitly check for `$y !== false` here?

File: inc/3rdparty/diff/Diff/Engine/Native.php [hunk 2]
```diff
@@ -196,18 +196,30 @@ protected function _diag ($xoff, $xlim, $yoff, $ylim, $nchunks)
...
                     if (empty($this->in_seq[$y])) {
                         $k = $this->_lcsPos($y);
                         assert($k > 0);
                         $ymids[$k] = $ymids[$k - 1];
                         break;
-                    } elseif ($y > $this->seq[$k - 1]) {
+                    }
+                    next($matches);
+                }
+                while ($y = current($matches)) {
```
Comment (euantorano) [inline inc/3rdparty/diff/Diff/Engine/Native.php]: Again, shouldn't we explicitly check `$y !== false`?

File: inc/3rdparty/diff/Diff/Engine/Native.php (hunk 1 above)
Comment (dvz) [reply inc/3rdparty/diff/Diff/Engine/Native.php]: Indeed; cc @johnbillion @SergeyBiryukov

**Comment by euantorano** [comment]
Looks good to me, but haven't tested yet.

**Comment by Stefan-ST** [comment]
Works for me.

**Pull request merged**